cpp/src/arrow/flight/server.cc

```diff
@@ -275,78 +277,137 @@ Status FlightServerBase::GetSchema(const ServerCallContext& context,
-        return GetNextDictionary(payload);
+    // Check if writer is already initialized or if schema has changed.
+    // To recreate the writer.
+    if (!writer_ || !batch->schema()->Equals(*reader_->schema())) {
+      if (writer_) {
+        RETURN_NOT_OK(writer_->Close());
+      }
+
+      // Create new writer with the batch's schema
+      auto payload_writer =
+          std::make_unique<ServerRecordBatchPayloadWriter>(&payload_list_);
+      auto writer_result = ipc::internal::OpenRecordBatchWriter(
+          std::move(payload_writer), batch->schema(), options_);
+
+      if (!writer_result.ok()) {
+        return writer_result.status();
+      }
+      writer_ = std::move(writer_result).ValueOrDie();
+      if (!payload_list_.empty()) {
+        // Drop Schema message if it was generated.
+        // We want new Dictionary or RecordBatch messages only.
+        payload_list_.pop_front();
       }
     }
 
-    RETURN_NOT_OK(reader_->ReadNext(&current_batch_));
+    // One WriteRecordBatch call might generate multiple payloads, so we
+    // need to collect them in a list.
+    RETURN_NOT_OK(writer_->WriteRecordBatch(*batch));
 
-    // TODO(ARROW-10787): Delta dictionaries
-    if (!current_batch_) {
-      // Signal that iteration is over
-      payload->ipc_message.metadata = nullptr;
+    // Return the first generated payload after WriteRecordBatch. There must be
+    // at least one payload generated.
+    if (!payload_list_.empty()) {
+      *payload = std::move(payload_list_.front());
+      payload_list_.pop_front();
       return Status::OK();
-    } else {
-      return ipc::GetRecordBatchPayload(*current_batch_, ipc_options_,
-                                        &payload->ipc_message);
     }
+
+    return Status::UnknownError("IPC writer didn't produce any payloads");
   }
 
-  Status Close() { return reader_->Close(); }
+  Status Close() {
+    if (writer_) {
+      RETURN_NOT_OK(writer_->Close());
+    }
+    return reader_->Close();
+  }
 
  private:
-  Status GetNextDictionary(FlightPayload* payload) {
-    const auto& it = dictionaries_[dictionary_index_++];
-    return ipc::GetDictionaryPayload(it.first, it.second, ipc_options_,
-                                     &payload->ipc_message);
-  }
+  // Simple payload writer that uses a list of generated payloads.
+  class ServerRecordBatchPayloadWriter : public ipc::internal::IpcPayloadWriter {
+   public:
+    explicit ServerRecordBatchPayloadWriter(std::list<FlightPayload>* payload_list)
+        : payload_list_(payload_list), first_payload_(true) {}
+
+    Status Start() override { return Status::OK(); }
+
+    Status WritePayload(const ipc::IpcPayload& ipc_payload) override {
+      FlightPayload payload;
+      payload.ipc_message = ipc_payload;
+
+      if (first_payload_) {
+        if (ipc_payload.type != ipc::MessageType::SCHEMA) {
+          return Status::Invalid("First IPC message should be schema");
+        }
+        first_payload_ = false;
+      }
```

Comment:
Do we really need to track first_payload_ and this, or just trust that the IPC writer does its job?

Reply:
Probably not, the IPC writer should guarantee that.